extensions/VisualEditor/wikia/modules/ve/ui/dialogs/ve.ui.WikiaInfoboxInsertDialog.js

```diff
@@ -155,29 +139,120 @@ ve.ui.WikiaInfoboxInsertDialog.prototype.insertInfoboxTemplate = function () {
 };
 
 /**
+ * @desc creates infobox item option wiget
+ * @param {Object} data
+ * @returns {OO.ui.DecoratedOptionWidget}
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.createInfoboxItemOptionWidget = function ( data ) {
+	return new OO.ui.DecoratedOptionWidget( {
+		data: data.title,
+		label:  data.title
+	} );
+};
+
+/**
+ * @desc creates select widget with indobox option widgets
+ * @param {OO.ui.DecoratedOptionWidget[]} items
+ * @returns {OO.ui.OoUiSelectWidget}
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.createInfoboxSelectWidget = function ( items ) {
+	var select = new OO.ui.SelectWidget();
+
+	select.addItems( items );
+
+	return select;
+};
+
+/**
  * Add the infobox template names to the dialog's SelectWidget
  *
  * @param {Object} data Response data from API
+ * @returns {Promise}
  */
-ve.ui.WikiaInfoboxInsertDialog.prototype.showResults = function ( data ) {
-	var i,
-		items = [],
-		infoboxes = data.query ? data.query.allinfoboxes : [];
-
-	if ( infoboxes.length > 0 ) {
-		for ( i = 0; i < infoboxes.length; i++ ) {
-			items.push(
-				new OO.ui.DecoratedOptionWidget({
-					data: infoboxes[i].title,
-					label:  infoboxes[i].title
-				})
-			);
-		}
-		this.select.addItems( items );
+ve.ui.WikiaInfoboxInsertDialog.prototype.createDialogContent = function ( data ) {
+	var deferred = $.Deferred(),
+		infoboxes = data.query ? data.query.allinfoboxes : [],
+		items = infoboxes.map( this.createInfoboxItemOptionWidget );
+
+	if ( items.length > 0 ) {
+		this.select = this.createInfoboxSelectWidget( items );
+
+		this.select.connect( this, {
+			select: 'onInfoboxTemplateSelect'
+		} );
+		deferred.resolve( this.select.$element );
+	} else {
+		this.getUnconvertedInfoboxes()
+			.then(this.createEmptyState)
+			.then(deferred.resolve)
 	}
+
+	return deferred.promise();
+};
+
+/**
+ * @desc gets list of unconverted infoboxes
+ * @returns {Promise}
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.getUnconvertedInfoboxes = function () {
+	var deferred = $.Deferred();
+
+	ve.init.target.constructor.static.apiRequest( {
+		action: 'query',
+		list: 'unconvertedinfoboxes'
+	} )
+		.done( function ( data ) {
+			deferred.resolve(
+				data.query &&
+				data.query.unconvertedinfoboxes &&
+				data.query.unconvertedinfoboxes.length > 0 ?
+					data.query.unconvertedinfoboxes :
+					[]
+			);
+		} )
+		.fail( function () {
+			deferred.resolve( [] );
+		} );
+
+	return deferred.promise();
+};
+
+/**
+ * @desc returns empty stage message
+ * @param {Array} unconvertedInfoboxes
+ * @returns {Promise}
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.createEmptyState = function ( unconvertedInfoboxes ) {
+	var deferred = $.Deferred(),
+		noInfoboxesHTML = '<span class="insert-infobox-empty-state">' +
```

Comment:
VisualEditor is based on OOJS and OOJS-UI which are widget oriented frameworks (as opposed to template oriented). Most likely the way to go here would be to create a new widget (which internally in its constructor, and not only, uses jQuery to create DOM) and use that widget here.
Please refrain from concatenating HTML like this - unless there is a really good reason for it (in some cases could be performance but not in this case).